src/qp/operators/ExternalSort.java

```diff
@@ -0,0 +1,279 @@
+/**
+ * Scans the base relational table
+ **/
+
+package qp.operators;
+
+import qp.utils.*;
+import java.io.*;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.PriorityQueue;
+
+import qp.optimizer.BufferManager;
+
+/**
+ * Scan operator - read data from a file
+ */
+public class ExternalSort extends Operator {
+	
+    boolean eos;           // To indicate whether end of stream reached or not
+    private ObjectInputStream sortedStream; // The input stream from which we read the sorted result.
+    
+    private ArrayList<Tuple> finalSortedTuples = new ArrayList<Tuple>();
+    
+    private Operator base;
+    private int numBuffer;
+    private Schema schema;
+    private int batchSize; // Number of tuples per outbatch;
+    private ArrayList<Attribute> projectList;
+    private int[] projectListIndices;
+    
+    public ExternalSort(Operator base, ArrayList<Attribute> projectList, int[] projectListIndices, int type, int batchSize) {
+        super(type);
+    	this.base = base;
+    	this.schema = base.getSchema();
+    	
+    	this.projectList=projectList;
+    	this.projectListIndices=projectListIndices;
+    	
+    	this.numBuffer = BufferManager.getTotalBuffers();
+    	this.batchSize = batchSize;
+    	this.eos = false;
+    
+    }
+    
+	/**
+     * Open file prepare a stream pointer to read input file
+     */
+    public boolean open() {
+    	// Phase 1: Created sorted runs
+    	int numOfSortedRuns = generateSortedRuns();
+    	// Phase 2: Merge sorted runs
+    	return mergeRuns(numOfSortedRuns, 1);
+    }
+
+    /**
+     * Next operator - get a tuple from the file
+     **/
+    public Batch next() {
```

Comment:
confusing name? so is this returning a tuple or Batch?

Reply:
returning one Batch of many tuples. (with an S)